esphome/components/ade7880/sensor.py

```diff
@@ -0,0 +1,261 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome.components import sensor, i2c
+from esphome import pins
+from esphome.const import (
+    CONF_ACTIVE_POWER,
+    CONF_APPARENT_POWER,
+    CONF_CALIBRATION,
+    CONF_CURRENT,
+    CONF_FORWARD_ACTIVE_ENERGY,
+    CONF_FREQUENCY,
+    CONF_ID,
+    CONF_NAME,
+    CONF_PHASE_A,
+    CONF_PHASE_ANGLE,
+    CONF_PHASE_B,
+    CONF_PHASE_C,
+    CONF_POWER,
+    CONF_POWER_FACTOR,
+    CONF_RESET_PIN,
+    CONF_REVERSE_ACTIVE_ENERGY,
+    CONF_VOLTAGE,
+    DEVICE_CLASS_APPARENT_POWER,
+    DEVICE_CLASS_CURRENT,
+    DEVICE_CLASS_ENERGY,
+    DEVICE_CLASS_POWER,
+    DEVICE_CLASS_POWER_FACTOR,
+    DEVICE_CLASS_VOLTAGE,
+    STATE_CLASS_MEASUREMENT,
+    STATE_CLASS_TOTAL_INCREASING,
+    UNIT_AMPERE,
+    UNIT_PERCENT,
+    UNIT_VOLT,
+    UNIT_VOLT_AMPS,
+    UNIT_VOLT_AMPS_REACTIVE_HOURS,
+    UNIT_WATT,
+    UNIT_WATT_HOURS,
+)
+
+DEPENDENCIES = ["i2c"]
+
+ade7880_ns = cg.esphome_ns.namespace("ade7880")
+ADE7880 = ade7880_ns.class_("ADE7880", cg.PollingComponent, i2c.I2CDevice)
+NeutralChannel = ade7880_ns.struct("NeutralChannel")
+PowerChannel = ade7880_ns.struct("PowerChannel")
+
+CONF_IRQ0_PIN = "irq0_pin"
+CONF_IRQ1_PIN = "irq1_pin"
+
+CONF_NEUTRAL = "neutral"
+
+NEUTRAL_CHANNEL_SCHEMA = cv.Schema(
+    {
+        cv.GenerateID(): cv.declare_id(NeutralChannel),
+        cv.Optional(CONF_NAME): cv.string_strict,
+        cv.Required(CONF_CURRENT): cv.maybe_simple_value(
+            sensor.sensor_schema(
+                unit_of_measurement=UNIT_AMPERE,
+                accuracy_decimals=2,
+                device_class=DEVICE_CLASS_CURRENT,
+                state_class=STATE_CLASS_MEASUREMENT,
+            ),
+            key=CONF_NAME,
+        ),
+        cv.Required(CONF_CALIBRATION): cv.Schema(
+            {
+                cv.Required(CONF_CURRENT): cv.int_,
```

Comment:
We try to enforce units on config entries so users know specifically what they are typing in.
Please apply this to the below calibration too.
```suggestion
                cv.Required(CONF_CURRENT): cv.All(cv.current, cv.int_),
```

Reply:
In this case there are no units; the value is a factor used in the calculations inside the chip. If this is a concern, I can make new CONF_ entries and avoid using CONF_CURRENT/VOLTAGE/etc.

Reply:
Perhaps a better name is `current_gain` etc?

Reply:
Sure, that will work. The datasheet says the registers are called "<X>GAIN" :-)

Reply:
Changes made; those three configuration elements are now `current_gain`, `power_gain`, and `voltage_gain`.